lib/Dialect/ArithExt/IR/ArithExtOps.td

```diff
@@ -0,0 +1,48 @@
+#ifndef LIB_DIALECT_ARITHEXT_IR_ARITHEXTOPS_TD_
+#define LIB_DIALECT_ARITHEXT_IR_ARITHEXTOPS_TD_
+
+include "lib/Dialect/ArithExt/IR/ArithExtDialect.td"
+include "mlir/IR/BuiltinAttributes.td"
+include "mlir/IR/CommonTypeConstraints.td"
+include "mlir/IR/OpBase.td"
+include "mlir/Interfaces/InferTypeOpInterface.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+
+
+class ArithExt_Op<string mnemonic, list<Trait> traits = [Pure]> :
+        Op<ArithExt_Dialect, mnemonic, traits> {
+  let cppNamespace = "::mlir::heir::arith_ext";
+}
+
+def ArithExt_BarrettReduceOp : ArithExt_Op<"barrett_reduce", [SameOperandsAndResultType]> {
+  let summary = "Compute the first step of the Barrett reduction.";
+  let description = [{
+    Let $q$ denote a statically known modulus and $b = 4^{w}$, where $w$ is the
+    smallest bit-width that contains the range $[0, q)$. The Barrett reduce
+    operation computes `barret_reduce x = x - floor(x * floor(b / q) / b) * q`.
+
+    Given $0 <= x < q^2$, then this will compute $(x \mod q)$ or $(x \mod q) + p$.
+  }];
+
+  let arguments = (ins
+    SignlessIntegerLike:$input,
+    I64Attr:$modulus
```

Comment:
I wonder if the cmod here should be more generic? In the verifier code above you take just the active bits, which is the right approach, and for now I doubt we would be performing mod q muls with greater than 64 bits, but just wanted to note.

Reply:
I think you're right that there are other assumptions sprinkled through the project in our lowerings. SG!